Riot/Managers/PushNotification/PushNotificationService.m

```diff
@@ -547,24 +547,36 @@ - (void)pushRegistry:(PKPushRegistry *)registry didReceiveIncomingPushWithPayloa
             //  when we have a VoIP push while the application is killed, session.callManager will not be ready yet. Configure it.
             [[AppDelegate theDelegate] configureCallManagerIfRequiredForSession:session];
             
-            if (lastCallInvite.isEncrypted)
-            {
-                [session decryptEvent:lastCallInvite inTimeline:nil];
-            }
-            
             NSLog(@"[PushNotificationService] didReceiveIncomingPushWithPayload: lastCallInvite: %@", lastCallInvite);
             
             if ([lastCallInvite.eventId isEqualToString:eventId])
             {
+                dispatch_group_t dispatchGroup = dispatch_group_create();
+                
+                dispatch_group_enter(dispatchGroup);
+                [session handleBackgroundSyncCacheIfRequiredWithCompletion:^{
+                    dispatch_group_leave(dispatchGroup);
+                }];
+                
+                if (lastCallInvite.isEncrypted && ![session decryptEvent:lastCallInvite inTimeline:nil])
```

Comment:
I think this code (and the full block) should be in the `handleBackgroundSyncCacheIfRequiredWithCompletion` completion block.
`decryptEvent` will return false if keys are not yet transferred from bg sync caches to the MXSession stores.

If we do that, we will probably not need to call `[session.callManager handleCallEvent:lastCallInvite];`.

I may be wrong but I do not understand the usage of a `dispatch_group_t` here and the fact we use it for one step but not all.

Reply:
They're intentionally not in the completion block, because we should report the new call in the same run loop, which means we cannot do that async.

Handling to-device events seems sync (so i think `decryptEvent` should succeed), but handling timeline events is async, so i think it's still better to call `[session.callManager handleCallEvent:lastCallInvite]`. 

We also want to continue to sync after processing the cache, for data integrity. We're using `dispatch_group_t` to only wait for the completion.

Reply:
Got it. Thanks. I failed to understand because I thought the completion block allowed to be fully async :/
Can you add this information in comments, please?

Reply:
Yes sure

Reply:
Comments added.